server/server/src/billing/payment_service.rs

```diff
@@ -0,0 +1,449 @@
+        .map_err(|e| VerificationError(format!("Could not verify stripe webhook: {}", e)))?;
+
+    let mut transaction = match state.index_db_client.begin().await {
+        Ok(t) => t,
+        Err(e) => {
+            return Err(StripeWebhookError::InternalError(format!(
+                "Cannot begin transaction: {:?}",
+                e
+            )))
+        }
+    };
+
+    match (event_type, &webhook.data.object) {
+        (StripeEventType::InvoicePaymentFailed, StripeObjectType::Invoice(invoice)) => {
+            match invoice.billing_reason {
+                StripeBillingReason::SubscriptionCycle => {
+                    file_index_repo::set_data_cap_with_stripe_customer_id(
+                        &mut transaction,
+                        &invoice.customer_id,
+                        FREE_TIER_SIZE,
+                    )
+                    .await
+                    .map_err(|e| {
+                        StripeWebhookError::InternalError(format!(
+                            "Cannot change data cap with customer id in Postgres: {:?}",
+                            e
+                        ))
+                    })?
+                }
+                _ => return Ok(()),
+            }
+        }
+        (StripeEventType::InvoicePaid, StripeObjectType::Invoice(partial_invoice)) => {
+            match partial_invoice.billing_reason {
+                StripeBillingReason::SubscriptionCycle => {
+                    let invoice = stripe_client::retrieve_invoice(&state, &partial_invoice.id)
+                        .await
+                        .map_err(|e| {
+                            StripeWebhookError::InternalError(format!(
+                                "Cannot retrieve expanded invoice: {:?}",
+                                e
+                            ))
+                        })?;
+
+                    let subscription = match invoice.subscription {
+                        StripeMaybeContainer::Unexpected(_) => {
+                            return Err(StripeWebhookError::InternalError(format!(
+                                "There should be a subscription tied to this invoice: {:?}",
+                                invoice
+                            )))
+                        }
+                        StripeMaybeContainer::Expected(subscription) => subscription,
+                    };
+
+                    file_index_repo::update_stripe_subscription_period_end(
+                        &mut transaction,
+                        &subscription.id,
+                        subscription.current_period_end,
+                    )
+                    .await
+                    .map_err(|e| {
+                        StripeWebhookError::InternalError(format!(
+                            "Couldn't update subscription period in Postgres: {:?}",
+                            e
+                        ))
+                    })?;
+                }
```

Comment:
I actually think we not only want to save the end date, but we want to create a table for every successful bill so a person can view their "transaction history" because we can't just email them an invoice. But we are going to redis-ify before that so don't bother until we talk.

Reply:
Are we still going with this, or will this be in a future pr?